controllers/pwa.js

```diff
@@ -163,30 +160,141 @@ router.post('/add', (req, res, next) => {
+          err.status = 404;
+          return next(err);
+        });
+    });
+});
+
+/**
+ * Generate the HTML with 'pwas/view.hbs' for one PWA
+ */
+function renderOnePwa(req, res, pwaId, contentOnly, url) {
+  return pwaLib.find(pwaId)
     .then(pwa => {
-      lighthouseLib.findByPwaId(req.params.pwa)
-      .then(lighthouse => {
-        let arg = Object.assign(libMetadata.fromRequest(req), {
-          pwa: pwa,
-          lighthouse: lighthouse,
-          rawManifestJson: JSON.parse(pwa.manifest.raw),
-          title: 'PWA Directory: ' + pwa.name,
-          description: 'PWA Directory: ' + pwa.name + ' - ' + pwa.description,
-          backlink: true,
-          contentOnly: contentOnly
+      return lighthouseLib.findByPwaId(pwaId)
+        .then(lighthouse => {
+          let arg = Object.assign(libMetadata.fromRequest(req, url), {
+            pwa: pwa,
+            lighthouse: lighthouse,
+            rawManifestJson: JSON.parse(pwa.manifest.raw),
+            title: 'PWA Directory: ' + pwa.name,
+            description: 'PWA Directory: ' + pwa.name + ' - ' + pwa.description,
+            backlink: true,
+            contentOnly: contentOnly
+          });
+          return render(res, 'pwas/view.hbs', arg);
         });
-        res.render('pwas/view.hbs', arg);
+    });
+}
+
+/**
+ * Generate the HTML with 'pwas/list.hbs' for a list of PWAs
+ */
+function renderPwaListPage(req, res) {
+  const pageNumber = parseInt(req.query.page, 10) || DEFAULT_PAGE_NUMBER;
+  const sortOrder = req.query.sort || DEFAULT_SORT_ORDER;
+  const start = parseInt(req.query.start, 10) || (pageNumber - 1) * LIST_PAGE_SIZE;
+  const limit = parseInt(req.query.limit, 10) || LIST_PAGE_SIZE;
+  const end = pageNumber * LIST_PAGE_SIZE;
+  const contentOnly = false || req.query.contentOnly;
+  let pwaCount = 0;
+  return pwaLib.count()
+    .then(count => {
+      pwaCount = count;
+      return pwaLib.list(start, limit, sortOrder);
+    })
+    .then(result => {
+      cachePwasFromList(req, res, result.pwas);
+      let arg = Object.assign(libMetadata.fromRequest(req), {
+        title: 'PWA Directory',
+        description: 'PWA Directory: A Directory of Progressive Web Apps',
+        pwas: result.pwas,
+        hasNextPage: result.hasMore,
+        hasPreviousPage: pageNumber > 1,
+        nextPageNumber: pageNumber + 1,
+        previousPageNumber: (pageNumber === 2) ? false : pageNumber - 1,
+        currentPageNumber: pageNumber,
+        sortOrder: (sortOrder === DEFAULT_SORT_ORDER) ? false : sortOrder,
+        showNewest: sortOrder === 'newest',
+        showScore: sortOrder === 'score',
+        pwaCount: pwaCount,
+        startPwa: start + 1,
+        endPwa: Math.min(pwaCount, end),
+        mainPage: true,
+        contentOnly: contentOnly
       });
+      return render(res, 'pwas/list.hbs', arg);
+    });
+}
+
+/**
+ * Cache individual PWAs from the list page if not in cache already.
+ */
+function cachePwasFromList(req, res, pwas) {
```

Comment:
There's also the problem of rendering with / without the shell. Rendering with the shell will only help users who are arriving directly to the page. Rendering without the shell will only help users navigating from the list. But since we are creating the cache on what's being rendered on the list, this shouldn't have much impact on the first case.